### ISSUE TITLE
Aborted requests for useQuery gives loading: true when it should give false

### ISSUE THREAD

**Issue by kumar935** [apollographql/apollo-client#9583]
I'm using AbortController to cancel queries/mutations and running into this issue where the loading state for these cancelled requests is `true` when I was expecting it to be `false`

**Intended outcome:**

The loading state from `useQuery` hook for cancelled request should be `false`

**Actual outcome:**

The loading state from `useQuery` hook for cancelled request should be `true`

**How to reproduce the issue:**

For @apollo/client 3.4.13: https://codesandbox.io/s/apollo-3-playground-3-4-13-9bmpcy 
For @apollo/client 3.5.10: https://codesandbox.io/s/apollo-3-playground-3-5-x-ryhg3x

1. Open any of the above sandboxes
2. Open the console panel where I am logging the loading states
3. Click on `refetch` button, it will initiate a call and immediately cancel it
4. Check loading state in console

**Versions**

3.4.13, 3.5.10

**Comment by tk04** [comment]
I see that you're using the useRef hook, and mutating it's .current attribute, which will not cause the component to re-render and update the useQuery state variables. 

_"The only difference between useRef() and creating a {current: ...} object yourself is that useRef will give you the same ref object on every render.

Keep in mind that useRef doesn’t notify you when its content changes. Mutating the .current property doesn’t cause a re-render. If you want to run some code when React attaches or detaches a ref to a DOM node, you may want to use a [callback ref](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node)[](https://reactjs.org/docs/hooks-reference.html#useimperativehandle) instead."_ 
[https://reactjs.org/docs/hooks-reference.html#useref](url)

One work around is to replace the useRef hook w/ useState: 

`import React, { useRef, useState } from "react";
import {
  ApolloProvider,
  ApolloClient,
  InMemoryCache,
  gql,
  useQuery,
  NetworkStatus
} from "@apollo/client";

const client = new ApolloClient({
  uri: "https://countries.trevorblades.com/",
  cache: new InMemoryCache()
});

const QUERY = gql`
  query GetContinents($input: [String]) {
    continents(filter: { code: { in: $input } }) {
      code
      name
    }
  }
`;

function Query() {
  const [aborterRef, setAbortRef] = useState(new AbortController());
  const { data, error, loading, networkStatus, refetch } = useQuery(QUERY, {
    variables: {
      input: "EU"
    },
    fetchPolicy: "network-only",
    context: {
      fetchOptions: {
        signal: aborterRef.signal
      }
    },
    notifyOnNetworkStatusChange: true
  });
  console.log("loading", loading);
  return (
    <>
      <h1>{networkStatus}</h1>
      <p>{JSON.stringify(data)}</p>
      <button
        onClick={() => {
          refetch({ input: "AF" });
          setTimeout(() => {
            aborterRef.abort();
            setAbortRef(new AbortController());
          }, 50);
        }}
      >
        refetch
      </button>
    </>
  );
}

export default function App() {
  return (
    <ApolloProvider client={client}>
      <div className="App">
        <h1>Hello CodeSandbox</h1>
        <Query />
      </div>
    </ApolloProvider>
  );
}
`
Hopefully this works for you.

**Comment by kumar935** [comment]
Thanks a lot @tk04 the workaround works for my case 👍

**Comment by kumar935** [comment]
@tk04 shouldn't the apollo hook trigger the re-render on its own though? Your workaround works for me for simple cases, but for more complicated cases doesn't seem to be convenient to trigger a re-render separately as your workaround suggests.

**Comment by tk04** [comment]
The Apollo hook will trigger a re-render only when the options you pass into it change. In your case, when you're assigning a new value to aborterRef.current, Apollo isn't aware of this change and thus doesn't trigger a re-render. One better work around is to call refetch after assigning a AbortController. For example, you can try this:

```
....
      <button
        onClick={() => {
          refetch({ input: "AF" });
          setTimeout(() => {
            aborterRef.current.abort();
            aborterRef.current = new AbortController();
            refetch(); // refetch after setting new AbortController
          }, 50);
        }}
      >

```

**Comment by kumar935** [comment]
Makes sense that apollo is not aware of aborterRef.current. But the end result is that the loading state is turned false. I am not aware of how it internally works, but when a request gets 200 or 500 response, for that scenario also we don't have to trigger refetch explicitly (or re-render the component explicitly), we just get the new loading state and component re renders, why doesn't the hook work in a similar way for a canceled request?

**Comment by tk04** [comment]
Good question, but I'm not really sure as why this is the case. I'm guessing that Apollo isn't aware that the request is even canceled, but I suggest you look into the useQuery source code.

**Comment by labithiotis** [comment]
I have a workaround that should work for anyone needing to abort a mutation, it will create a new abortController for each new request and when aborted will set the request loading to false. It may not be perfect but should work for now.

```tsx
import { useCallback, useRef, useState } from 'react';
import { ApolloCache, DefaultContext, MutationOptions, OperationVariables, useApolloClient } from '@apollo/client';

export function useAbortableMutation<
  TData = any,
  TVariables extends OperationVariables = OperationVariables,
  TContext extends Record<string, any> = DefaultContext,
  TCache extends ApolloCache<any> = ApolloCache<any>,
>(
  mutation: any,
): [
  (options: Omit<MutationOptions<TData, TVariables, TContext>, 'mutation'>) => Promise<void>,
  { data?: TData | null; error?: Error; loading: boolean },
  () => void,
] {
  const client = useApolloClient();
  const controller = useRef(new window.AbortController());
  const [data, setData] = useState<TData | null>();
  const [error, setError] = useState<Error>();
  const [loading, setLoading] = useState(false);

  const request = useCallback(
    async (options: Omit<MutationOptions<TData, TVariables, TContext>, 'mutation'>) => {
      try {
        setLoading(true);
        controller.current = new window.AbortController();
        const res = await client.mutate<TData, TVariables, TContext, TCache>({
          ...options,
          mutation,
          context: { ...options.context, fetchOptions: { signal: controller.current.signal } } as unknown as TContext,
          errorPolicy: 'all',
        });
        setData(res.data);
      } catch (error) {
        if (error instanceof Error) {
          setError(error);
        }
      } finally {
        setLoading(false);
      }
    },
    [client, mutation],
  );

  const abort = useCallback(() => {
    if (!controller.current.signal.aborted) {
      controller.current.abort();
      setLoading(false);
    }
  }, [controller]);

  return [request, { data, error, loading }, abort];
}
```

```tsx
export function() {
  const  [doSomething, { loading, error, data }, abort] = useAbortableMutation<Variables, Data>(gql`mutation DoSomething() { doSomething() }`);  
}
```

**Comment by jeel-toddleapp** [comment]
Actually, if you check code for httpLink, abortError is not simply swallowed. Don't know the reasoning behind it

**Comment by jerelmiller** [comment]
@jeel-toddleapp we didn't either 😆. This behavior will actually be changing in the upcoming v3.8 release and should fix this issue (for reference, here is the work #11058). In 3.8, if you pass a `signal` and then abort that signal, it will propagate that abort error, which should also fix the loading state. If you'd like to try it out, install the latest `rc` release: 

```
npm i @apollo/client@rc
```

**Comment by jerelmiller** [comment]
I have gone ahead and tried the latest `rc` out on the original reproduction: https://codesandbox.io/s/apollo-3-playground-3-4-13-forked-fmc9cn?file=/src/App.js

Since the original issue appears fixed by #11058, I'm going to go ahead and close this issue. Thanks!
